srcjs/output_binding_html.js

```diff
@@ -107,23 +107,48 @@ function renderDependency(dep) {
   }
 
   if (dep.stylesheet) {
-    var stylesheets = $.map(asArray(dep.stylesheet), function(stylesheet) {
-      var this_href = href + "/" + encodeURI(stylesheet);
-      // If a styleSheet with this href is already active, then disable it
-      var re = new RegExp(this_href + "$");
-      for (var i = 0; i < document.styleSheets.length; i++) {
-        var h = document.styleSheets[i].href;
-        if (!h) continue;
-        if (restyle && h.match(re)) document.styleSheets[i].disabled = true;
+    var links = $.map(asArray(dep.stylesheet), function(stylesheet) {
+      var link = $("<link rel='stylesheet' type='text/css'>");
+      var this_href =  href + "/" + encodeURI(stylesheet);
+      if (!restyle) {
+        return link.attr("href", this_href);
       }
-      // Force client to re-request the stylesheet if we've already seen it
-      // (without this unique param, the request might get cached)
-      if (restyle) {
-        this_href = this_href + "?restyle=" + new Date().getTime();
+      // When we're restyling (i.e., this is coming through
+      // session$setCurrentTheme()), then before rendering the new
+      // stylesheet, search the existing styleSheets for href(s) that
+      // appear to point to match the same source, and flag for removal
+      for (var i = 0; i < document.styleSheets.length; i++) {
+        var sheet = document.styleSheets[i];
+        if (!sheet.href) continue;
+        var pieces = sheet.href.split("?")[0].split("/");
+        if (pieces[pieces.length - 1] === encodeURI(stylesheet) &&
+            pieces[pieces.length - 2] === href) {
+          sheet._shiny_restyle = true;
+        }
       }
-      return $("<link rel='stylesheet' type='text/css'>").attr("href", this_href);
+      // Include a timestamp to force the client to re-request 'new'
+      // stylesheets (otherwise it might get cached)
+      var param = "?restyle=" + new Date().getTime();
+      return link.attr("href", this_href + param);
     });
-    $head.append(stylesheets);
+
+    $head.append(links);
+
+    // Disable and remove old stylesheets on the next tick
+    // The timeout amount here is currently a total guess...
+    // At first I had 10 milliseconds, but encountered some FOUC
+    // with bootstraplib's real-time themer. In the future, if we encounter
+    // other examples of FOUC we may want to consider increasing the timeout
+    if (restyle) {
+      setTimeout(function() {
+        for (var i = 0; i < document.styleSheets.length; i++) {
```

Comment:
I see a potential issue here with calling `sheet.ownerNode.remove()` while iterating over `document.styleSheets`. I think there's a good chance this will alter `document.styleSheets` as you're iterating over it, which is not a good idea, since that means that this loop could skip over elements. (That's why I had suggested adding a callback that took a separate final pass after all the sheets were disabled, and removed any elements that had been disabled.)

Also, is `document.styleSheets` OK to use? See:
https://caniuse.com/mdn-api_documentorshadowroot_stylesheets
https://developer.mozilla.org/en-US/docs/Web/API/DocumentOrShadowRoot/styleSheets.